src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_CAPACITY = 10;
+
+    private int size;
+
+    private Object[] dataArray;
+
+    public ArrayList() {
+        this.dataArray = new Object[START_CAPACITY];
+        this.size = 0;
+    }
+
+    private boolean sizeCheck() {
+        return size == dataArray.length;
+    }
+
+    private void grow(int minCapacity) {
+        int oldCapacity = dataArray.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity < minCapacity) {
+            newCapacity = minCapacity;
+        }
+        dataArray = Arrays.copyOf(dataArray, newCapacity);
+    }
+
     @Override
     public void add(T value) {
-
+        if (sizeCheck()) {
+            grow(size + 1);
+        }
+        dataArray[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
```

Comment:
The index check logic is repeated in multiple methods (`add(T value, int index)`, `get(int index)`, `set(T value, int index)`, and `remove(int index)`). Consider extracting this logic into a separate private method to avoid code duplication.